_posts/2015-09-28-service-discovery.md

```diff
@@ -0,0 +1,57 @@
+>or any of the Consul agents. The agents forward queries to the servers automatically.
+
+For more details, see the full [introduction](https://www.consul.io/intro/index.html)  and the [in-depth architecture overview](https://www.consul.io/docs/internals/architecture.html) on the Consul web site.
+
+##Health-Checks
+The data inside the service registry only provides real value if it is up to date. Just knowing there once was a service available at a certain location is not enough. Therefore, checking the availability or health status of all service instances is crucial for service discovery.
+
+In our setup, the health checking is also done by Consul. The servers check the general availability of the agents, and therefore the hosts. In addition, you can register different types of service checks with an agent, which then perform these checks against the services registered locally in fixed intervals. Whenever such a check fails, the according service or host will be marked as unavailable, and therefore the service registry will no longer promote the faulty instance any longer.
+
+##Lookup / Routing
+The third major building block is a component, which is able to look up or route requests to an appropriate service instance, given an identifier for the target service. With this identifier, the clients should either receive an actual service instance address to connect to, or have its requests routed transparently to the service instance.
+
+We decided for the routing solution, to keep the logic and also library dependencies out of the services. As the identifier, we use the first path element of any request URI, which means that a request to `http://insert.hostname.here/myservice/recource1` will be routed to an appropriate service instance of the `myservice` service.
+
+With this naming schema defined, the actual routing is done by [HAProxy](http://www.haproxy.org/), a reliable, high performance TCP/HTTP load balancer. It is configured by a process called consul-template, which queries the information about available service instances from Consul and applies it to a provided template, in this case a template HAProxy configuration file, and (gracefully) restarts HAProxy afterwards.
+
+{% image blog-consul-template-haproxy.png %}{% endimage %}
+
+With this in place, HAProxy routes any incoming request to the most appropriate instance of the service given by the first path element of the request. the determination of the most appropriate instance of the service can be decided by the various load balancing features of HAProxy.
+
+##Dynamic Service Registration
+So far, service discovery works fine, but how do the services register with the Consul agent on the hosts on which they are started? To solve this, we introduce another component, [Registrator](http://gliderlabs.com/registrator/latest/). This is a so-called service registry bridge for Docker. What is actually does is nicely summarized on their website:
```

Comment:
What is actually does is nicely summarized on their website > What it actually does is nicely summarised on their website